server/controllers/authController.js

```diff
@@ -181,45 +183,94 @@ This link will expire in 24 hours.`)
       .catch(err => res.status(500).json({message: errorMessage(err)}));
   };
 
-  this.loginUser = async function (req, res) {
-    const errors = validationResult(req);
-    if (!errors.isEmpty()) {
-      res.status(400).json({
-        message: "Invalid request",
-        errors: errors.array()
-      });
-      return;
-    }
-    // Check the credentials
-    let user;
-    try {
-      user = await userRepository.getByEmail(req.body.email);
-    } catch (err) {
-      res.status(500).json({message: errorMessage(err)});
-      return;
-    }
-    if (!user || !validatePassword(req.body.password, user.password)) {
-      res.status(400).json({message: "Invalid username/password"});
-      return;
-    }
-    const lastLogin = user.lastLogin;
-    const currentDate = moment();
-    await userRepository.update(user, {
-      lastLogin: currentDate
-    })
-      .then(() => {
-        res.status(200).json({
-          message: "Successful login!",
-          user: {
-            uid: user.id,
-            isAdmin: user.isAdmin,
-            lastLogin: lastLogin,
-            token: generateToken(user)
-          }
-        });
+    this.loginUser = async function (req, res) {
+      const errors = validationResult(req);
+      if (!errors.isEmpty()) {
+        res.status(400).json({message: "Invalid request", errors: errors.array()});
+        return;
+      }
+      // Check there is a user
+      let user;
+      try {
+        user = await userRepository.getByEmail(req.body.email);
+      } catch (err) {
+        res.status(500).json({message: errorMessage(err)});
+        return;
+      }
+      if (!user) {
+        res.status(400).json({message: "Invalid Username/Password. Please try again."});
+        return;
+      }
+      // Check if account is locked
+      const currentDate = moment();
+      if (user.unlockDate && currentDate.isBefore(user.unlockDate)) {
+        res.status(400).json({message: "Invalid Username/Password. Please try again."});
+        return;
+      }
+
+      // Check password
+      if (!validatePassword(req.body.password, user.password)) {
+        const newPasswordRetries = user.passwordRetries - 1;
+        if (newPasswordRetries <= 0) {
+          // Lock account if that was last try
+          // Restore password retries so when account unlocks they can try again
+          const unlockDate = moment().add(10, 'minutes').format();
+          await userRepository.update(user, {
```

Comment:
I don't see why this is a problem. We do this in all functions now for consistency.
When we are detecting client errors we make those checks synchronously. Then once we've identified the steps we need to make to complete the request then we we can chain those steps together (using thens) and just wait on the overall result. Otherwise you have to wrap every single request in a try catch, when we only really use try catch because it will likely be followed by an early return.